medschool/src/parse.rs

```diff
@@ -0,0 +1,312 @@
+/// Converts a [`syn::ItemEnum`] into a [`PartialType`].
+/// Currently, we don't handle enum variants as fields, we just use its top-level docs.
+impl TryFrom<syn::ItemEnum> for PartialType<'_> {
+    type Error = ();
+
+    fn try_from(item: syn::ItemEnum) -> Result<Self, Self::Error> {
+        let ident = item.ident.to_string();
+        Ok(PartialType {
+            ident: ident.into(),
+            docs: docs_from_attributes(item.attrs).ok_or(())?,
+            fields: Default::default(),
+        })
+    }
+}
+
+/// Converts a [`syn::ItemStruct`] into a [`PartialType`].
+/// We remove docs that have `<!--${internal}-->` (they're not converted to [`PartialType`]s).
+impl TryFrom<syn::ItemStruct> for PartialType<'_> {
+    type Error = ();
+
+    fn try_from(item: syn::ItemStruct) -> Result<Self, Self::Error> {
+        let fields = item
+            .fields
+            .into_iter()
+            .filter_map(|field| PartialField::try_from(field).ok())
+            .collect::<BTreeSet<_>>();
+
+        let ident = item.ident.to_string();
+        Ok(PartialType {
+            ident: ident.into(),
+            docs: docs_from_attributes(item.attrs).ok_or(())?,
+            fields,
+        })
+    }
+}
+
+/// Converts a [`syn::Item`] into a [`PartialType`], if possible.
+impl TryFrom<syn::Item> for PartialType<'_> {
+    type Error = ();
+
+    fn try_from(item: syn::Item) -> Result<Self, Self::Error> {
+        match item {
+            syn::Item::Struct(item_struct) => PartialType::try_from(item_struct),
+            syn::Item::Enum(item_enum) => PartialType::try_from(item_enum),
+            syn::Item::Mod(item_mod) => PartialType::try_from(item_mod),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Converts a list of [`syn::File`] into a [`BTreeSet`] of our own [`PartialType`] types, so we can
+/// get a root node (see the [`Ord`] implementation of `PartialType`).
+#[tracing::instrument(level = "trace", ret)]
+pub fn parse_docs_into_set<'a>(
+    files: Vec<syn::File>,
+) -> Result<HashSet<PartialType<'a>>, DocsError> {
+    let type_docs = files
+        .into_iter()
+        // go through each `File` extracting the types into a hierarchical tree based on which types
+        // belong to other types
+        .flat_map(|syntaxed_file| {
+            syntaxed_file
+                .items
+                .into_iter()
+                // convert an `Item` into a `PartialType`
+                .filter_map(|item| PartialType::try_from(item).ok())
+        })
+        .collect::<HashSet<_>>();
+
+    Ok(type_docs)
+}
+
+/// DFS helper function to resolve the references of the types. Returns the docs of the fields of
+/// the field we're currently looking at search till its leaf nodes. The leaf here means a primitive
+/// type for which we don't have a [`PartialType`].
+fn dfs_fields<'a>(
+    field: &PartialField<'a>,
+    types: &'a HashSet<PartialType>,
+    cache: &mut HashMap<&'a str, Vec<String>>,
+    recursion_level: &mut usize,
```

Comment:
I think we need a cap for this, to avoid crashing if the recursion goes too deep.

Probably a `const` on the function that we check to abort.